### PR TITLE
[SofaPython] PythonScriptDataEngine (PSDE)

### PR DESCRIPTION
**Pull request by sescaida**
The PSDE  works similar to a function block in Simulink or comparable software. The user can define a block (a PSDE) declaring its Inputs, Outputs (I/Os) and its behavior using Python alone. It has some similarities with the "PythonScriptController" (PSC). The two main differences are:

- PSDE-objects are not triggered by events (not conceptually at least). Rather, the functionality should be implemented in the update() function of the associated script, which is called once per simulation frame.  
- I/Os can be added directly to the object inside the parse() function of the PSDE-object. Linking using the '@'-syntax also works. Therefore, the PSDE-object behaves like any other object that has I/Os (or Data in general). Inside the script these newly defined I/Os can be accessed as members of the object, providing an easy syntax.

You can think of the PSDE as a possibility of subclassing the Sofa 'DataEngine' in a scripted way. In the second example provided it is shown how a versatile ROI functionality can be implemented with the help of the PSDE.

Two examples with some documentation are provided in:  
applications/plugins/SofaPython/examples/PSDEExamples/

Example of use (01-basic/PSDEBasic.pyscn):
```python
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import Sofa
import os

path = os.path.dirname(os.path.abspath(__file__))+'/../../../../../../share/mesh/'
        
def createScene(rootNode):
        rootNode.createObject("MeshVTKLoader", name="loader", filename=path+'liver.vtk')
        rootNode.createObject('PythonScriptDataEngine', name='PSDE', filename="DataEngine.py", classname="MyDataEngine")	       

```
Example for DataEngine class in Python (01-basic/DataEngine.py):

```python
class MyDataEngine(Sofa.PythonScriptDataEngine):           
  
    def parse(self):        
        # this is the correct place to define inputs and outputs! (linking will not work in init() for instance)
        self.addNewInput('Positions',datatype='p',value='@loader.position')       
        self.addNewOutput('NumPoints',datatype='d', value='0')                

    def init(self):
	pass

    def update(self):        
        self.NumPoints = len(self.Positions)
```

What it does NOT provide: A new form of parsing from Python data to Sofa data. It uses the available mechanism that uses string conversion and parsing. Also, the type of the Data has to be declared/selected with a parameter in Python, because there is no logic to automatically derive the type desired by the user (could be future work).

______________________________________________________

This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

File: applications/plugins/SofaPython/Binding_Base.cpp [hunk 1]
```diff
@@ -35,84 +36,353 @@ using namespace sofa::core::objectmodel;
...
+        return nullptr;
+    }
+
+    BaseData* bd = nullptr ;
+//    bool OldSchoolParsing = false; //TODO (Stefan Escaida 29.01.2018): should a direct parsing mechanism Python-->Sofa be implemented? (instead of going through str)
+    if(dataRawType[0] == 's'){
+        Data<std::string>* t = new Data<std::string>() ;
+        t = new(t) Data<std::string>(obj->initData(t, std::string(""), dataName, dataHelp)) ;
+        bd = t;        
+    }
```
Comment (ErwanDouaille) [inline applications/plugins/SofaPython/Binding_Base.cpp]: A factory would be beter :)
Something like : BaseData* bf = factory(dataRawType[0]);

Comment (ErwanDouaille) [reply applications/plugins/SofaPython/Binding_Base.cpp]: Here is an example of a factory with sofa : https://github.com/SofaDefrost/sofa/blob/sofaCommunication/applications/plugins/Communication/components/serverCommunicationOSC.inl#L50

File: CMakeLists.txt [hunk 2]
```diff
@@ -323,7 +323,7 @@ sofa_add_plugin( applications/plugins/SofaPython SofaPython )
 option(SOFA_BUILD_SCENECREATOR "Compile the C++ API located in applications/plugins/SceneCreator" OFF)
 
 ## Tutorials option
-option(SOFA_BUILD_TUTORIALS "Build (most of) the tutorials available." OFF)
+option(SOFA_BUILD_TUTORIALS "Build (most of) the tutorials available." ON)
```
Comment (guparan) [inline CMakeLists.txt]: I'm not sure you really wanted to add this change in your PR. If yes, why?

File: CMakeLists.txt [hunk 3]
```diff
@@ -366,6 +366,8 @@ if(SOFA_BUILD_TUTORIALS)
     add_subdirectory(applications/tutorials)
 endif()
 
+#add_subdirectory(applications/tutorials)
```
Comment (guparan) [inline CMakeLists.txt]: ... same remark ;-)

File: CMakeLists.txt [hunk 4]
```diff
@@ -366,6 +366,8 @@ if(SOFA_BUILD_TUTORIALS)
...
+
```
Comment (guparan) [inline CMakeLists.txt]: ... same remark ;-)

**Comment by guparan** [comment]
[ci-build]

File: CMakeLists.txt (hunk 4 above)
Comment (sescaida) [reply CMakeLists.txt]: Yes, you're right, it slipped in in both cases!

File: applications/plugins/SofaPython/Binding_Base.cpp (hunk 1 above)
Comment (sescaida) [reply applications/plugins/SofaPython/Binding_Base.cpp]: Ok, thanks! I will look into it.

File: applications/plugins/SofaPython/Binding_Base.cpp [hunk 5]
```diff
@@ -37,13 +37,30 @@
...
 using sofa::core::topology::BaseMeshTopology ;
-using sofa::helper::vector ;
-using namespace sofa::core::objectmodel;
+typedef BaseMeshTopology::Tetra Tetra;
+using sofa::helper::vector;
 using sofa::helper::Factory;
+using namespace sofa::core::objectmodel;
 
-
+// TODO (Stefan Escaida 13.02.2018): this factory code is redundant to the Communication plugin, but should easily be mergeable, when an adequate spot is found.
```
Comment (ErwanDouaille) [inline applications/plugins/SofaPython/Binding_Base.cpp]: Nice ! What did you plan for the redundant part ?

Comment (sescaida) [reply applications/plugins/SofaPython/Binding_Base.cpp]: Eventually, one and the same factory could be used for creating the Data for Sofa Communication and for the PSDE. Merging the code into one should be trivial. Yet, as far as I understand, currently there is no obvious place in the structure where this factory should be --> future work?

Comment (ErwanDouaille) [reply applications/plugins/SofaPython/Binding_Base.cpp]: Not sure than merging code into one is consistent. They are both doing something similar but it's a simple instanciation of a factory class.

**Comment by damienmarchal** [comment]
Args, shouldn't this one should be in to-review status ?

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests]

**Comment by sescaida** [comment]
Hi guys, I made the changes concerning your feedback and now the PR is ready for review from my side! 

The main changes are:
- Created a basic example of usage
- Now uses a factory for data creation
- The use of both args and kwargs for I/O creation is possible
- No unnecessary changes in files / no geometries added, etc. (cleaning)

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests]

**Comment by damienmarchal** [comment]
@sescaida the CI is reporting there is new warning added by this PR (not a lot). Can you fix them before we merge ?

**Comment by bcarrez** [comment]
[ci-build][with-scene-tests]

**Comment by bcarrez** [comment]
[ci-build][with-scene-tests]

**Pull request merged**